File: WindowsApplication7/MainForm.vb
Imports System.Net
Imports System.IO
Imports System.Text
Imports System.Text.RegularExpressions

Public Class MainForm

    Private Async Sub getHtmlButton_Click(sender As Object, e As EventArgs) Handles getHtmlButton.Click
        Dim inputBuffer(1024000) As Byte

        getHtmlButton.Enabled = False
        Dim webRequest As HttpWebRequest = Await StartRequestAsync()
        Dim index As Integer = 0
        Using response As WebResponse = Await webRequest.GetResponseAsync()
            Dim stream As Stream = response.GetResponseStream()
            Dim bytesRead As Integer
            Do
                bytesRead = Await stream.ReadAsync(inputBuffer, index, inputBuffer.Length - index)
                index += bytesRead
            Loop Until bytesRead = 0

            Trace.WriteLine(String.Format("Read {0} bytes", index))
        End Using
        Dim html As String = Encoding.ASCII.GetString(inputBuffer, 0, index)

        listBox.Items.Clear()
        If (Not String.IsNullOrEmpty(html)) Then
            For Each x As String In GetScriptBodies(html)
                listBox.Items.Add(x)
            Next
        End If
        getHtmlButton.Enabled = True

    End Sub

    Private Function StartRequestAsync() As Task(Of HttpWebRequest)
        Return Task.Factory.StartNew(Function() DirectCast(Net.WebRequest.Create("http://google.ca"), HttpWebRequest))
    End Function

    Private Shared Iterator Function GetScriptBodies(html As String) As IEnumerable(Of String)
        Const pattern As String = "<script.*?>\s*(?'scriptBody'.+?)\s*</script>"
        Dim matches As MatchCollection = Regex.Matches(html, pattern)
        For Each match As Match In matches
            Dim scriptBody As String = match.Groups("scriptBody").Value
            Yield scriptBody
        Next
    End Function

End Class

File: WindowsFormsApplication7b/MainForm.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace WindowsFormsApplication7b
{
    public partial class MainForm : Form
    {
        private CancellationTokenSource cancellationTokenSource;
        private CancellationToken cancellationToken;

        public MainForm()
        {
            InitializeComponent();
        }

        private class MyProgress : Progress<int>
        {
            public void Progress(int value)
            {
                OnReport(value);
            }
        }

        private async void getHtmlButton_Click(object sender, EventArgs e)
        {
            cancellationTokenSource = new CancellationTokenSource();
            cancellationToken = cancellationTokenSource.Token;

            getHtmlButton.Enabled = false;
            progressBar.Value = 0;
            progressBar.Visible = true;
            cancelButton.Enabled = true;
            cancelButton.Visible = true;

            HttpWebRequest webRequest = await StartRequestAsync();

            byte[] inputBuffer;
            int index = 0;
            using (WebResponse response = await webRequest.GetResponseAsync())
            {
                inputBuffer = new byte[response.ContentLength];

                Stream stream = response.GetResponseStream();

                MyProgress progress = new MyProgress();
                progress.ProgressChanged += progress_ProgressChanged;

                int bytesRead;
                do
                {
                    try
                    {
                        bytesRead = await stream.
                                              ReadAsync(inputBuffer, index,
                                                        inputBuffer.Length - index,
                                                        cancellationToken);
                        if (cancellationToken.IsCancellationRequested) break;
                        index += bytesRead;
                        progress.Progress((index*100)/inputBuffer.Length);
                    }
                    catch (TaskCanceledException)
                    {
                        break;
                    }
                } while (bytesRead != 0);
            }

            listBox.Items.Clear();
            if (!cancellationToken.IsCancellationRequested)
            {
                Trace.WriteLine(String.Format("Read {0} bytes", index));
                string html = Encoding.ASCII.GetString(inputBuffer, 0, index);
                if (!String.IsNullOrEmpty(html))
                {
                    foreach (var x in GetScriptBodies(html))
                        listBox.Items.Add(x);
                }
            }
            else
            {
                listBox.Items.Add("CANCELLED");
                progressBar.Visible = false;
            }
            getHtmlButton.Enabled = true;
            cancelButton.Enabled = false;
            cancelButton.Visible = false;
        }

        void progress_ProgressChanged(object sender, int e)
        {
            progressBar.Value = e;
        }

        private static Task<HttpWebRequest> StartRequestAsync()
        {
            return
                Task.Factory.
                    StartNew(
                        () =>
                        (HttpWebRequest)
                        WebRequest.Create("http://en.wikipedia.org/wiki/Line_of_succession_to_the_British_throne"));
        }

        private static IEnumerable<string> GetScriptBodies(string html)
        {
            const string pattern = @"<script.*?>\s*(?'scriptBody'.+?)\s*</script>";
            MatchCollection matches = Regex.Matches(html, pattern);
            foreach (Match match in matches)
            {
                string scriptBody = match.Groups["scriptBody"].Value;
                yield return scriptBody;
            }
            yield break;
        }

        private void cancelButton_Click(object sender, EventArgs e)
        {
            cancellationTokenSource.Cancel();
        }
    }
}
